Fix symlinking of plugins / ensure plugins know their non-canonicalized paths

fixpunkt opened the pull request:
- This fixes resource resolution when plugins are symlinked
- For a detailed explanation of the base issue, see #24

fixpunkt committed:
Pass non-canonicalized path to Plugin::__construct()

This fixes resource resolution when plugins are symlinked

fixpunkt committed:
Compute non-canonicalized plugin class path using path arithmetic

mitelg commented on src/Core/Kernel.php:
```diff
@@ -346,7 +346,7 @@ private function instantiatePlugins(array $plugins): void
 
             /** @var Plugin $plugin */
-            $plugin = new $className((bool) $pluginData['active']);
+            $plugin = new $className((bool) $pluginData['active'], $this->getProjectDir() . $pluginData['path']);
```
please consider that the path to the plugin root and the path to the plugin base class could be different. so this is not correct solution for the symlink use case :wink:

fixpunkt replied:
Supporting symlinks while still resolving all resources relative to the Plugin's base class in a clean way is likely impossible, since the path of a class in PHP is unfortunately always canonicalized (which is the root cause of all the symlinking problems). I have a somewhat hacky solution on this branch: https://github.com/shopware/platform/compare/master...VIISON:pickware/experimental-compute-non-canonicalized-plugin-class-path#diff-b3226395e77d614094c66a2d3202f556R67

But as you can see, that's not pretty either.

I've also discovered that resolving resources relative to the plugin base class breaks layouts that look like this:

```
.
├── Resources
│   └── administration
│       └── main.js
└── src
    └── MyPlugin.php
```

Because apparently, overriding `\Shopware\Core\Framework\Bundle::getAdministrationEntryPath` with `return '../Resources/administration'` doesn't work either.

Also, with the relative-to-base-class model, the values for the `plugin.path` column and the value of  `Plugin->getPath()` differ, which feels kind of inconsistent.

I think resolving resources relative to the plugin root is the more logical and consistent option and causes less problems than resolving relative to the base class. If resolving relative to the base class is required, it should be trivial to override the `get*Path()` methods in the plugin to achieve this - the opposite case unfortunately doesn't work.

I've also already explained the non-standard layout problem shown by the filesystem tree above to @janbuecker.

All in all, I'd like to petition you to reconsider dropping the relative-to-base-class resolution in favor of resolving resources relative to the plugin root.

If that's not a workable solution for you, we'd have to use the hacky workaround on the https://github.com/shopware/platform/compare/master...VIISON:pickware/experimental-compute-non-canonicalized-plugin-class-path branch or drop symlink support, which would be pretty inconvenient for plugin developers. Also, I have no idea on how to make the `/Resources` case described above work correctly with that approach.

fixpunkt commented:
I've done some more work on this. [Here's the set of plugins](https://github.com/shopware/platform/files/3072380/AdministrationExamplePlugin.zip) I symlinked into `custom/plugins` to test some approaches.

The approach I ended up implementing was to use path arithmetic to compute the path containing the plugin bundle path in terms of the (symlinked) plugin base path under `custom/plugins`, as opposed to using the canonical form return by `Bundle::getPath()` which breaks resource resolution for symlinked plugins. I've also added `Plugin::getPluginPath()` for Plugins which need to know their plugin base path as opposed to the Plugin class' containing directory.

The drawback of this approach is that the layout used by `AdministrationExamplePlugin3` is not supported:

```
.
├── Resources
│   ├── administration
│   │   └── main.js
│   └── views
│       └── administration
│           └── index.html.twig
├── composer.json
└── src
    └── AdministrationExamplePlugin3.php
```

Note that the current plugin system on `master` also doesn't support this.

I've tried a more involved approach that could theoretically support this use case here: https://github.com/shopware/platform/compare/master...VIISON:pickware/INCOMPLETE-alternative-fix-symlinked-plugins-again

This almost works, except for the fact that Symfony's `assets:install` command hard-codes where it looks for `Resources/public`: https://github.com/symfony/framework-bundle/blob/72cce34ae1feba0249b4a2be253fd8aabaaf6100/Command/AssetsInstallCommand.php#L140 😕

So yeah, I think the code now in this PR is the best compromise which doesn't break the existing recommended plugin layout.

fixpunkt commented:
Also note: If the plan is to `composer require` plugins using a path repository with `symlink: true` in the future, that means symlinked plugins must be fully supported.

janbuecker commented:
Thanks @fixpunkt, can you provide a unit test for your new path resolution?

fixpunkt committed:
Add fallback to fix existing tests

fixpunkt committed:
Use new-style plugin layout for unit test plugin

Signed-off-by: Jannik Jochem <jannik.jochem@pickware.com>

fixpunkt committed:
Add unit test for path resolution logic, refactor

fixpunkt committed:
Merge branch 'master' into pickware/fix-symlinked-plugins-again

mitelg commented on src/Core/Kernel.php:
```diff
@@ -351,7 +351,7 @@ private function instantiatePlugins(array $plugins): void
 
             /** @var Plugin $plugin */
-            $plugin = new $className((bool) $pluginData['active']);
+            $plugin = new $className((bool) $pluginData['active'], $this->getProjectDir() . $pluginData['path']);
```
a slash is missing between the pathes, because of this commit: https://github.com/shopware/platform/commit/b36982eb3241066e9ff7f30956d86ea3d8cfd7ac

fixpunkt replied:
Thanks for the heads-up!

Is there a way to test this part of the logic?

fixpunkt committed:
Add missing slash

fixpunkt commented:
Good point about the tests @janbuecker. I've also fixed the tests I broke with the previous changes without noticing, sorry.

One thing that still bothers me is that I cannot make the `$basePath` parameter of `Plugin::__construct()` required because I'd have to change the property sort order, which is a bit of a larger change. But forcing every client to pass in the plugin's base path explicitly would probably save us all a lot of symlink-related headaches in the future.

I'd really like to make both of the constructor parameters required since having a default probably only makes testing marginally easier, but adds a lot of ambiguity. What is your opinion on this?

mitelg commented:
@fixpunkt I have no problem with making the path required. I made it optional to have the possible fallback to the symfony bundle logic. 
@janbuecker ?

janbuecker commented:
As long as a plugin developer does not have to care about it, I'd be fine with the change if it would make life easier using symlinks.

fixpunkt commented:
Cool, I'll add some more commits tomorrow!

fixpunkt committed:
Merge branch 'master' into pickware/fix-symlinked-plugins-again

fixpunkt committed:
Make Plugin::__construct() arguments required

fixpunkt commented:
I've updated the PR as discussed, added extra tests which assert that symlinked plugins are handled correctly and ensured all other tests pass.

This PR is ready for the next round of review.

fixpunkt committed:
Merge branch 'master' into pickware/fix-symlinked-plugins-again

janbuecker commented:
Thank you @fixpunkt. I've created an internal ticket to resolve this PR soon. You can track the status here: https://issues.shopware.com/issues/NEXT-3679
